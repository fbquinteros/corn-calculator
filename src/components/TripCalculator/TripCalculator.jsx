import React, { useState, useEffect } from "react";
import "./style/index.css";

const TripCalculator = ({ initialCornQuantity, initialGeeseQuantity }) => {
  // Put consts or functions or state
  const [cornQuantity, setCornQuantity] = useState(initialCornQuantity || 0);
  const [geeseQuantity, setGeeseQuantity] = useState(initialGeeseQuantity || 0);
  const [message, setMessage] = useState("No cargo to take");
  const [tripCount, setTripCount] = useState(1);
  const costPerTrip = 0.25;

  useEffect(() => {
    calculateMessage();
  }, [cornQuantity, geeseQuantity, initialGeeseQuantity, initialCornQuantity]);

  useEffect(() => {
    if (cornQuantity === 1 && geeseQuantity === 1) {
      return setTripCount(cornQuantity + geeseQuantity);
    }

    if (cornQuantity >= 1 && geeseQuantity >= 1) {
      return setTripCount(0);
    }

    if (cornQuantity <= 0 && geeseQuantity <= 0) {
      return setTripCount(1);
    }
    return setTripCount(Math.max(cornQuantity, 0) + Math.max(geeseQuantity, 0));
  }, [cornQuantity, geeseQuantity, initialGeeseQuantity, initialCornQuantity]);
  
  const calculateMessage = () => {
    if (cornQuantity === 0 && geeseQuantity === 0) {
      return setMessage(`No cargo to take`);
    }

    if (
      (cornQuantity > 1 && geeseQuantity > 0) ||
      (cornQuantity > 0 && geeseQuantity > 1)
    ) {
      return setMessage(`Do not make trip, corn is at risk`);
    }

    if (geeseQuantity === cornQuantity && geeseQuantity >= 1) {
      return setMessage(`take the corn first`);
    }

    if (geeseQuantity > 0) {
      return setMessage(
        `take in this order : ${geeseQuantity === 1 ? "goose" : "geese"}`
      );
    }

    if (cornQuantity > 0) {
      return setMessage(
        `take in this order :  ${cornQuantity === 1 ? "corn" : "corns"}`
      );
    }
  };
  const calculateCostOfTrip = () => {
    if (
      (cornQuantity > 1 && geeseQuantity > 0) ||
      (cornQuantity > 0 && geeseQuantity > 1)
    ) {
      return parseFloat(0).toFixed(2);
    }

    if (cornQuantity > 0) {
      return parseFloat(cornQuantity * costPerTrip).toFixed(2);
    }

    if (geeseQuantity > 0) {
      return parseFloat(geeseQuantity * costPerTrip).toFixed(2);
    }

    if (cornQuantity <= 0 && geeseQuantity <= 0) {
      return parseFloat(costPerTrip).toFixed(2);
    }
  };

  // DRY
  const cornQuantityHandler = (e) => {
    let value = Number.parseInt(e.target.value, 10) || 0;
    console.log("Quantity of corn", value);
    if (value < 0) {
      value = 0;
    }
    setCornQuantity(value);
  };

  // DRY
  const geeseQuantityHandler = (e) => {
    let value = Number.parseInt(e.target.value, 10) || 0;
    console.log("Quantity of geese", value);
    if (value < 0) {
      value = 0;
    }
    setGeeseQuantity(value);
  };

  return (
    <form>
      <legend>Plan your trip</legend>
      {/* TODO: Extract Trip Calculator Component */}
      <fieldset>
        <div className="input-group">
          {/* TODO: Extract Quantity Input Component */}
          <label forhtml="cornQuantity">Corn Quantity:</label>
          <input
            type="number"
            id="cornQuantity"
            name="quantity"
            value={cornQuantity}
            onChange={cornQuantityHandler}
            min={0}
          />
        </div>
        <div className="input-group">
          {/* TODO: Extract Quantity Input Component */}
          <label forhtml="geeseQuantity">Geese Quantity:</label>
          <input
            type="number"
            id="geeseQuantity"
            name="quantity"
            value={geeseQuantity}
            onChange={geeseQuantityHandler}
            min={0}
          />
        </div>
      </fieldset>
      <label>Â£ {calculateCostOfTrip()}</label>
      <p className="user-info">{message}</p>
      <p className="user-info">Trip Count: {tripCount}</p>
    </form>
  );
};

export default TripCalculator;
