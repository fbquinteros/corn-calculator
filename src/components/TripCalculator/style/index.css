.input-group  {
    margin-top:10px;
    width:60%;
    display:flex;
    justify-content:space-between;
    flex-wrap:wrap;
  }
  
  .user-info {
    font-weight: bold;
    font-style: italic;
  }
  
  form {
    margin:0 auto;
    padding:20px;
  }
  
  label, input {
    flex-basis:200px;
  }