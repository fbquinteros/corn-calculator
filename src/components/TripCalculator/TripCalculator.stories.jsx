import React, { Fragment } from "react";
import TripCalculator from "./TripCalculator";

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  title: "Component/TripCalculator",
  component: TripCalculator,
};

const TripCalculatorTemplate = (args) => (
  <Fragment>
    <TripCalculator {...args} />
  </Fragment>
);

export const DefaultZeros = TripCalculatorTemplate.bind({});
DefaultZeros.args = {
  initialCornQuantity: 0,
  initialGeeseQuantity: 0,
};

export const OneGooseAndZeroCorn = TripCalculatorTemplate.bind({});
OneGooseAndZeroCorn.args = {
  initialCornQuantity: 0,
  initialGeeseQuantity: 1,
};

export const TwoGooseAndZeroCorn = TripCalculatorTemplate.bind({});
TwoGooseAndZeroCorn.args = {
  initialCornQuantity: 0,
  initialGeeseQuantity: 2,
};

export const OneCornAndZeroGoose = TripCalculatorTemplate.bind({});
OneCornAndZeroGoose.args = {
  initialCornQuantity: 1,
  initialGeeseQuantity: 0,
};

export const TwoCornAndZeroGoose = TripCalculatorTemplate.bind({});
TwoCornAndZeroGoose.args = {
  initialCornQuantity: 2,
  initialGeeseQuantity: 0,
};

export const OneCornAndOneGoose = TripCalculatorTemplate.bind({});
OneCornAndOneGoose.args = {
  initialCornQuantity: 1,
  initialGeeseQuantity: 1,
};