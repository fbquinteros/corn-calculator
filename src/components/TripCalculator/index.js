
import TripCalculator from './TripCalculator';
export default TripCalculator;    
