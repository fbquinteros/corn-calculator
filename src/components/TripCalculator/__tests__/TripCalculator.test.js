import React from "react";
import { render } from "@testing-library/react";

import { DefaultZeros, OneGooseAndZeroCorn, OneCornAndZeroGoose } from "../TripCalculator.stories";

describe("Trip Calculator", () => {
  test("should default with 0 goose, 0 corn : cost £0.25 : no cargo to take", () => {
    const { container } = render(<DefaultZeros {...DefaultZeros.args} />);

    expect(container).toMatchSnapshot("default zeros");
  });

  test("should get 1 goose, 0 corn : cost £0.25 : take in this order : goose", () => {
    const { container } = render(<OneGooseAndZeroCorn {...OneGooseAndZeroCorn.args} />);

    expect(container).toMatchSnapshot("one goose only");
  });

  test("should get 0 goose, 1 corn : cost £0.25 : take in this order : corn", () => {
    const { container } = render(<OneCornAndZeroGoose {...OneCornAndZeroGoose.args} />);

    expect(container).toMatchSnapshot("one corn only");
  });
});
