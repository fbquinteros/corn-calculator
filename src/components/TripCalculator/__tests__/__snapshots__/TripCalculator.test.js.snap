// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trip Calculator should default with 0 goose, 0 corn : cost £0.25 : no cargo to take: default zeros 1`] = `
<div>
  <form>
    <legend>
      Plan your trip
    </legend>
    <fieldset>
      <div
        class="input-group"
      >
        <label
          forhtml="cornQuantity"
        >
          Corn Quantity:
        </label>
        <input
          id="cornQuantity"
          name="quantity"
          type="number"
          value="0"
        />
      </div>
      <div
        class="input-group"
      >
        <label
          forhtml="geeseQuantity"
        >
          Geese Quantity:
        </label>
        <input
          id="geeseQuantity"
          name="quantity"
          type="number"
          value="0"
        />
      </div>
    </fieldset>
    <label>
      £ 
      0.25
    </label>
    <p
      class="user-info"
    >
      No cargo to take
    </p>
  </form>
</div>
`;

exports[`Trip Calculator should get 0 goose, 1 corn : cost £0.25 : take in this order : corn: one corn only 1`] = `
<div>
  <form>
    <legend>
      Plan your trip
    </legend>
    <fieldset>
      <div
        class="input-group"
      >
        <label
          forhtml="cornQuantity"
        >
          Corn Quantity:
        </label>
        <input
          id="cornQuantity"
          name="quantity"
          type="number"
          value="1"
        />
      </div>
      <div
        class="input-group"
      >
        <label
          forhtml="geeseQuantity"
        >
          Geese Quantity:
        </label>
        <input
          id="geeseQuantity"
          name="quantity"
          type="number"
          value="0"
        />
      </div>
    </fieldset>
    <label>
      £ 
      0.25
    </label>
    <p
      class="user-info"
    >
      take in this order :  corn
    </p>
  </form>
</div>
`;

exports[`Trip Calculator should get 0 goose, 2 corn : cost £0.50 : take in this order : corns: two corn only 1`] = `
<div>
  <form>
    <legend>
      Plan your trip
    </legend>
    <fieldset>
      <div
        class="input-group"
      >
        <label
          forhtml="cornQuantity"
        >
          Corn Quantity:
        </label>
        <input
          id="cornQuantity"
          name="quantity"
          type="number"
          value="2"
        />
      </div>
      <div
        class="input-group"
      >
        <label
          forhtml="geeseQuantity"
        >
          Geese Quantity:
        </label>
        <input
          id="geeseQuantity"
          name="quantity"
          type="number"
          value="0"
        />
      </div>
    </fieldset>
    <label>
      £ 
      0.50
    </label>
    <p
      class="user-info"
    >
      take in this order :  corns
    </p>
  </form>
</div>
`;

exports[`Trip Calculator should get 1 goose, 0 corn : cost £0.25 : take in this order : goose: one goose only 1`] = `
<div>
  <form>
    <legend>
      Plan your trip
    </legend>
    <fieldset>
      <div
        class="input-group"
      >
        <label
          forhtml="cornQuantity"
        >
          Corn Quantity:
        </label>
        <input
          id="cornQuantity"
          name="quantity"
          type="number"
          value="0"
        />
      </div>
      <div
        class="input-group"
      >
        <label
          forhtml="geeseQuantity"
        >
          Geese Quantity:
        </label>
        <input
          id="geeseQuantity"
          name="quantity"
          type="number"
          value="1"
        />
      </div>
    </fieldset>
    <label>
      £ 
      0.25
    </label>
    <p
      class="user-info"
    >
      take in this order : goose
    </p>
  </form>
</div>
`;

exports[`Trip Calculator should get 1 goose, 1 corn : cost £0.50 : take corn first: one of each 1`] = `
<div>
  <form>
    <legend>
      Plan your trip
    </legend>
    <fieldset>
      <div
        class="input-group"
      >
        <label
          forhtml="cornQuantity"
        >
          Corn Quantity:
        </label>
        <input
          id="cornQuantity"
          name="quantity"
          type="number"
          value="1"
        />
      </div>
      <div
        class="input-group"
      >
        <label
          forhtml="geeseQuantity"
        >
          Geese Quantity:
        </label>
        <input
          id="geeseQuantity"
          name="quantity"
          type="number"
          value="1"
        />
      </div>
    </fieldset>
    <label>
      £ 
      0.25
    </label>
    <p
      class="user-info"
    >
      take the corn first
    </p>
  </form>
</div>
`;

exports[`Trip Calculator should get 1 goose, 2 corn : Do not make trip, corn is at risk: one goose and 2 corns 1`] = `
<div>
  <form>
    <legend>
      Plan your trip
    </legend>
    <fieldset>
      <div
        class="input-group"
      >
        <label
          forhtml="cornQuantity"
        >
          Corn Quantity:
        </label>
        <input
          id="cornQuantity"
          name="quantity"
          type="number"
          value="2"
        />
      </div>
      <div
        class="input-group"
      >
        <label
          forhtml="geeseQuantity"
        >
          Geese Quantity:
        </label>
        <input
          id="geeseQuantity"
          name="quantity"
          type="number"
          value="1"
        />
      </div>
    </fieldset>
    <label>
      £ 
      0.00
    </label>
    <p
      class="user-info"
    >
      Do not make trip, corn is at risk
    </p>
  </form>
</div>
`;
